docs/DafnyRef/integration-java/IntegrationJava.md

```diff
@@ -0,0 +1,215 @@
+place theses scripts on the system `$PATH`.
+
+Alternately, one can use `dafny translate` to create the `.java` translation of the Dafny program and then use regular Java build processes to combine the
+Dafny-generated Java code with other Java code.
+
+## **Matching Dafny and Java types**
+
+If the Java method has input arguments or an output value, then the Dafny declaration must use
+corresponding types in the Dafny declaration:
+
+|-------------------------------|-----------------------------|
+|  Dafny type                   |   Java type                 |
+|-------------------------------|-----------------------------|
+| bool                          | boolean                     |
+| int                           | java.math.BigInteger        |
+| int64                         | long                        |
+| int32                         | int                         |
+| int16                         | short                       |
+| int8                          | byte                        |
+| char                          | char                        |
+| bitvector                     | appropriate int-based type  |
+| ORDINAL                       | java.math.BigInteger        |
+| real                          | dafny.BigRational           |
+|                               | double                      |
+|                               | float                       |
+| string                        | dafny.DafnySequence<? extends Character>  |
+| JavaString                    | java.lang.String                        |
+| C, C? (for class, iterator C) | (class) C                   |
+| (trait) T                     | (interface) T                |
+| datatype, codatatype          | (class) C                   |
+| subset type                   | same as base type           |
+| tuple                         | dafny.Tuple_n_              |
+| array<T>                      | T'[]                        |
+| seq<T>                        | dafny.DafnySequence<? extends T'> |
+| set<T>, iset<T>               | dafny.DafnySet<? extends T'>      |
+| multiset<T>                | dafny.DafnySet<? extends T'>      |
+| map<T,U>, imap<T,U>           | dafny.DafnyMap<? extends T'>      |
+| imap<T,U>, imap<T,U>           | dafny.DafnyMap<? extends T'>      |
+| function (arrow) types        | java.util.function.Function<T',U'> |
+|-------------------------------|------------------------------|
+
+Either there is a direct match between the Dafny-generated Java types and the types used in the Java code (as listed in the table above), or some wrapper
+code is needed either on the Java side or on the Dafny side.
+
+- If Dafny is calling a Java method in some Java library and no alterations
+to the Java library are possible or desired, then the Dafny code must be written
+in such a way that the Dafny types will translate directly to the Java types.
```

Comment:
It would be good to have one example on how to transform a string in one way or the other, or how to build a dafny sequence.

Reply:
Agreed. But that requires some library functions -- so I put off explaining this until I could do a bit more investigation.